File: OJTrain/Go/src/leetcode2.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	l3 := &ListNode{}
	l3.Next = nil
	l3.Val = 0
	current := l3
	carry := 0

	for l1 != nil && l2 != nil {
		node := ListNode{}
		node.Next = nil
		node.Val = 0
		current.Next = &node
		current = current.Next
		sum := l1.Val + l2.Val + carry
		current.Val = sum % 10
		carry = sum / 10
		l1 = l1.Next
		l2 = l2.Next
	}

	for l1 != nil {
		node := ListNode{}
		node.Next = nil
		node.Val = 0
		current.Next = &node
		current = current.Next
		sum := l1.Val + carry
		current.Val = sum / 10
		carry = sum % 10
		l1 = l1.Next
	}

	for l2 != nil {
		node := ListNode{}
		node.Next = nil
		node.Val = 0
		current.Next = &node
		current = current.Next
		sum := l2.Val + carry
		current.Val = sum / 10
		carry = sum % 10
		l2 = l2.Next
	}

	if carry > 0 {
		current.Next = &ListNode{}
		current = current.Next
		current.Val = 1
	}
	return l3.Next
}

File: OJTrain/Python/leetcode2.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        carry = 0
        l3 = ListNode()
        current = l3
        while l1 and l2:
            current.next = ListNode()
            current = current.next
            num = l1.val + l2.val + carry
            current.val = num % 10
            carry = int(num / 10)
            l1 = l1.next
            l2 = l2.next

        while l1:
            current.next = ListNode()
            current = current.next
            num = l1.val + carry
            current.val = num % 10
            carry = int(num / 10)
            l1 = l1.next

        while l2:
            current.next = ListNode()
            current = current.next
            num = l2.val + carry
            current.val = num % 10
            carry = int(num / 10)
            l2 = l2.next

        if carry > 0:
            current.next = ListNode()
            current = current.next
            current.val = 1
        return l3.next
